Fix ensure context binding of root object

illfang opened the pull request:
This PR is an addition to the commit 00c926a4cc66d5ae5ed0a4c25c7b933c358e7224 for solving #1354

In the previous commit, a new root object will be created during the `OnProcessMessageReceived` when there has not yet been a JS context created by the browser. By evanluating JS in the `OnProcessMessageReceived` function, a context will be created and `OnContextCreated` will be called.
As the root object has already been created the following will be the case:
```
LOG(WARNING) << "A context has been created for the same browser / frame without context released called previously";
```
This will skip the `Bind` of the root object to this new context which this PR tries to solve.

amaitland commented on CefSharp.BrowserSubprocess.Core/CefAppUnmanagedWrapper.cpp:
```diff
@@ -64,19 +64,22 @@ namespace CefSharp
+            rootObject = gcnew JavascriptRootObjectWrapper(browser->GetIdentifier(), browserWrapper->BrowserProcess);
+            rootObjectWrappers->TryAdd(frameId, rootObject);
+        }
+        if (rootObject->IsBound)
```
Formatting - Add line between `}` and `if`

amaitland commented on CefSharp.BrowserSubprocess.Core/JavascriptRootObjectWrapper.h:
```diff
@@ -40,6 +41,10 @@ namespace CefSharp
         {
             CefSharp::Internals::JavascriptCallbackRegistry^ get();
         }
+        property bool IsBound
```
Formatting - Add line between `}` and `property`

amaitland commented on CefSharp.BrowserSubprocess.Core/JavascriptRootObjectWrapper.cpp:
```diff
@@ -14,6 +14,11 @@ namespace CefSharp
+        {
+            throw gcnew InvalidOperationException("This root object has already been bound.");
+        }
+        _isBound = true;
```
Formatting - Add lines above and below

amaitland commented:
@illfang Do you have a use case for this? Or just adding support for the edge case?

amaitland commented:
Other than the formatting looks fine :+1:

illfang commented:
> Other than the formatting looks fine

Thank you. I've adjusted the formatting and updated the PR.


> Do you have a use case for this? Or just adding support for the edge case?

I've created two branches demonstrating the current problem:
```
Uncaught ReferenceError: testTime is not defined
@ undefined:1:21
```
https://github.com/illfang/CefSharp/tree/fix-ensure-context-binding-of-root-object-example

and one demonstrating the working fix:
```
991/7/23
```
https://github.com/illfang/CefSharp/tree/fix-ensure-context-binding-of-root-object-example-fixed

amaitland commented:
Thanks :+1: Merging into `cefsharp/43` as well.
